ENH: Implement the DLPack Array API protocols for ndarray.

hameerabbasi opened the pull request:
<!--         ----------------------------------------------------------------
                MAKE SURE YOUR PR GETS THE ATTENTION IT DESERVES!
                ----------------------------------------------------------------

*  FORMAT IT RIGHT:
      http://www.numpy.org/devdocs/dev/development_workflow.html#writing-the-commit-message

*  IF IT'S A NEW FEATURE OR API CHANGE, TEST THE WATERS:
      http://www.numpy.org/devdocs/dev/development_workflow.html#get-the-mailing-list-s-opinion

*  HIT ALL THE GUIDELINES:
      https://numpy.org/devdocs/dev/index.html#guidelines

*  WHAT TO DO IF WE HAVEN'T GOTTEN BACK TO YOU:
      http://www.numpy.org/devdocs/dev/development_workflow.html#getting-your-pr-reviewed
-->

Fixes #19013 

cc @rgommers @mattip @seberg

TODO
- [x] `from_dlpack` function.
- [x] dtype support in `__dlpack__`

TODO items added/edited by seberg:
- [x] Move all dlpack specific code to a single C file, rather than `methods.c`
- [ ] I would feel much better if there is a *PR to `dlpack.h`* that clarifies alignment requirements (and use of `byte_offset`).  After such a PR exists, we can review the current rules when to reject export.  (if the rejection is not guaranteed to be fixed by `arr.copy()` we may have a problem.)
- [ ] As discussed in the meeting `from_dlpack` shall not be exposed as new, public API.  There will be no public API to convert to NumPy _from_ `__dlpack__`.  It would be ok to have experimental/private, underscored API.
- [ ] (preferably) Make sure there is a PR to either `dlpack.h` or array-api that states that imported arrays shall be considered writable and exporters like NumPy should avoid exporting readonly arrays (truly readonly only?).  Ideally, with a footnote that this means that users of immutable array libraries may be up for a big surprise.
    (mattip says:) Not sure how to resolve this. The [data api issue](https://github.com/data-apis/array-api/issues/191) is still open
- [x] I am not quite sure if the `->deleter == NULL` case is settled (I _think_ it is, but maybe there needs to be some clarification in the header as well).
- [x] Specify cleanup to know for a fact whether this implementation or the CuPy implementation is correct.

hameerabbasi committed:
ENH: Add the __dlpack__ and __dlpack_device__ methods to ndarray.

rgommers commented:
Thanks Hameer, looks like a good start. 

Of the points @seberg brought up on the issue, adding a version attribute seems the most important - because checking for version will be needed before any backwards-incompatible change (such as adding an extra field) can be done. I asked about it on https://github.com/dmlc/dlpack/issues/34, and the suggestion was to add it as an attribute on the `PyCapsule`. Could you look into doing that?

BvB93 commented on numpy/core/src/multiarray/methods.c:
```diff
@@ -2694,6 +2696,152 @@ array_complex(PyArrayObject *self, PyObject *NPY_UNUSED(args))
+        return NULL;
+    }
+
+    if (stream != Py_None)
+    {
+        PyErr_SetString(PyExc_RuntimeError, "NumPy only supports stream=None.");
+        return NULL;
+    }
```
This might be more of a problem with the array-api docs, but [`__dlpack__`](https://data-apis.org/array-api/latest/API_specification/array_object.html#dlpack-self-stream-none) currently makes no mention of integer-support being optional for `stream`.

hameerabbasi replied:
Streams are irrelevant for CPU arrays, IIUC, so I make sure it's not supplied. Another decision could be to ignore it completely.

rgommers replied:
The parameter description is: `stream : Optional[ int ]`

Anything except `None` is clearly incorrect on CPU, so +1 for raising.

BvB93 replied:
I agree that raising is the correct decision here, but the docs currently suggest that `__dlpack__` implementations _must_ support both `None` and `int` for the `stream` parameter. 
In my opinion it's worth clarifying that support for `stream: int` is, in fact, optional.

rgommers replied:
Okay, just a "on CPU and other devices without streams, only None is accepted"?

BvB93 replied:
> Okay, just a "on CPU and other devices without streams, only None is accepted"?

Yeah, this sounds about right.

hameerabbasi commented on numpy/core/src/multiarray/methods.c:
```diff
@@ -2694,6 +2696,152 @@ array_complex(PyArrayObject *self, PyObject *NPY_UNUSED(args))
+    managed->manager_ctx = self;
+    managed->deleter = array_dlpack_deleter;
+
+    PyObject *capsule = PyCapsule_New(managed, NPY_DLPACK_CAPSULE_NAME, array_dlpack_capsule_deleter);
```
```suggestion
    PyObject *capsule = PyCapsule_New(managed, NPY_DLPACK_CAPSULE_NAME, NULL);
```

Unsure if the `PyCapsule` needs a deleter.

seberg replied:
Well, we have to be able to delete it to free the memory, but maybe we need some state as well, to ensure that we don't delete it twice if the consumer already deletes it without deleting the capsule (might be good for non-reference counted Python also?).

I thought I read that there was some capsule renaming involved (to signal that its "invalid" after deletion).  I guess setting the data to NULL or so would be just as well.  (If this is standardized, do that. If not setting data to NULL to guard against deleting twice? Once from python once from the consumer.)

hameerabbasi commented:
> Could you look into doing that?

Sure -- I commented there instead of here, so more of the relevant parties can see it.

BvB93 commented on numpy/core/src/multiarray/methods.c:
```diff
@@ -2694,6 +2696,152 @@ array_complex(PyArrayObject *self, PyObject *NPY_UNUSED(args))
+    else
+    {
+        PyErr_SetString(PyExc_TypeError,
+                        "DLPack only supports signed/unsigned integers, float and complex dtypes.");
```
Just to verify: DLPack doesn't support boolean dtypes?

hameerabbasi replied:
It could in theory, but it doesn't as of version 0.5.0.

BvB93 reviewed (commented):
Hi Hameer, could you also add these two new methods to the `ndarray` stubs in [`numpy/__init__.pyi`](https://github.com/numpy/numpy/blob/main/numpy/__init__.pyi)?
Their signatures are fairly simple, so that's a plus:
``` python
# NOTE: Every single import below is already present in `__init__.pyi`
# and doesn't have to be repeated
from typing import Any, Tuple
from typing_extensions import Literal as L

from numpy import number
from numpy.typing import NDArray

# `builtins.PyCapsule` unfortunately lacks annotations as of the moment;
# use `Any` as a stopgap measure
_PyCapsule = Any

def __dlpack__(self: NDArray[number[Any]], *, stream: None = ...) -> _PyCapsule: ...
def __dlpack_device__(self) -> Tuple[L[1], L[0]]: ...
```

seberg commented on numpy/core/src/multiarray/methods.c:
```diff
@@ -2694,6 +2696,152 @@ array_complex(PyArrayObject *self, PyObject *NPY_UNUSED(args))
+    
+    managed_dtype.bits = 8 * itemsize;
+    managed_dtype.lanes = 1;
+    if (PyDataType_ISSIGNED(dtype))
```
We need to also check for the DType being native.  I assume unaligned is OK?

seberg commented on numpy/core/src/multiarray/methods.c:
```diff
@@ -2694,6 +2696,152 @@ array_complex(PyArrayObject *self, PyObject *NPY_UNUSED(args))
+    {
+        managed_dtype.code = kDLUInt;
+    }
+    else if (PyDataType_ISFLOAT(dtype))
```
Maybe we should just explicitly list them?  We cannot include longdouble here, since the C99 standard doesn't specify it (I am not even sure IBM double-double is strictly standard conform).  And I assume `Float` in DLPack means IEEE compatible float format of defined size.  float16, float32, and float64 are fine, for the others maybe DLPack should add an extended float field or so, if we want it?

I just don't see longdoublel to be much good, considering its already system dependend on the CPU alone...

seberg commented on numpy/core/src/multiarray/methods.c:
```diff
@@ -2694,6 +2696,152 @@ array_complex(PyArrayObject *self, PyObject *NPY_UNUSED(args))
+    {
+        free(managed);
+        free(managed_shape);
+        free(managed_strides);
```
Would probably just do `if (capsule == NULL) { ... return ...}`

Introducing a `goto fail;` may make sense, but its probably not worthwhile.

seberg commented on numpy/core/src/multiarray/methods.c:
```diff
@@ -2694,6 +2696,152 @@ array_complex(PyArrayObject *self, PyObject *NPY_UNUSED(args))
+}
+
+static PyObject *
+array_dlpack_device(PyArrayObject *self, PyObject *args)
```
```suggestion
array_dlpack_device(PyArrayObject *NPY_UNUSED(self), PyObject *NPY_UNUSED(args))
```
although I am not sure the compilers actually complain...

seberg reviewed (commented):
Had started looking at it, so a few small comments.  It looks good, I have to think a bit more about the API, but that has nothing to do with the implementation.

I think if we put the header where you put it, it will be effectively public.  I guess that is fine thoug? Otherwise I think placing it into `common` is good.

BvB93 commented on numpy/__init__.pyi:
```diff
@@ -1644,6 +1644,10 @@ _ArrayComplex_co = NDArray[Union[bool_, integer[Any], floating[Any], complexfloa
 _ArrayNumber_co = NDArray[Union[bool_, number[Any]]]
 _ArrayTD64_co = NDArray[Union[bool_, integer[Any], timedelta64]]
 
+# `builtins.PyCapsule` unfortunately lacks annotations as of the moment;
+# use `Any` as a stopgap measure
+_PyCapsule = Any
+
```
Interestingly, if we could get `PyCapsule` in typeshed annotated as a parameterizable type,
then it would in principle be possible for static type checkers to read, _e.g._, the underlying shape and dtype.

The only thing that users or libraries would have to do here is declare the necessary annotations.

Perhaps something to consider for the future?

Examples
---------

``` python
from typing import TypeVar, Any, Generic, Tuple as Tuple
import numpy as np

# Improvised `PyCapsule` annotation
_T = TypeVar("_T")
class PyCapsule(Generic[_T]): ...

# Construct a more compact `PyCapsule` alias.; `Tuple` used herein to introduce 2 parameters 
# (there may be more appropriate types that can fulfill this function)
_Shape = TypeVar("_Shape", bound=Any)  # TODO: Wait for PEP 646's TypeVarTuple
_DType = TypeVar("_DType", bound=np.dtype[Any])
DLPackCapsule = PyCapsule[Tuple[_Shape, _DType]]

# A practical example
def from_dlpack(__x: DLPackCapsule[_Shape, _DType]) -> np.ndarray[_Shape, _DType]: ...
```

hameerabbasi replied:
I'll consider this as out of scope of this PR for now, but will leave the conversation unresolved for visibility.

hameerabbasi commented:
> This name should be some standardized DLPack name? Or is this intentional for now to keep it effectively private right now?

> Well, we have to be able to delete it to free the memory, but maybe we need some state as well, to ensure that we don't delete it twice if the consumer already deletes it without deleting the capsule (might be good for non-reference counted Python also?).
> 
> I thought I read that there was some capsule renaming involved (to signal that its "invalid" after deletion). I guess setting the data to NULL or so would be just as well. (If this is standardized, do that. If not setting data to NULL to guard against deleting twice? Once from python once from the consumer.)

There's a standardized name, see data-apis/array-api#186. I have modified the PR accordingly.

> We need to also check for the DType being native. I assume unaligned is OK?

Unaligned is OK, see https://github.com/numpy/numpy/issues/19013#issuecomment-843735344. How does one check for native dtypes?

> Maybe we should just explicitly list them? We cannot include longdouble here, since the C99 standard doesn't specify it (I am not even sure IBM double-double is strictly standard conform). And I assume Float in DLPack means IEEE compatible float format of defined size. float16, float32, and float64 are fine, for the others maybe DLPack should add an extended float field or so, if we want it?
> 
> I just don't see longdoublel to be much good, considering its already system dependend on the CPU alone...

Can we verify this?

hameerabbasi commented:
> Can we verify this?

Ignore, it does mean IEEE float. Does `np.float80` or `np.float128` not respect the IEEE convention? In that case, it makes sense to limit both complex and floats.

eric-wieser commented:
`np.float80` etc are all just aliases for `np.longdouble` based on what C produces for `sizeof(long double)`

seberg commented:
I guess if longdouble is 64bit, you can be sure its just double. That we could check for if we want.  There are some C macros, but I am not sure any of them check for IEEE 128bit double. AFAIK it is rare enough at the time that I am not sure I care right now.

Technically, most current hardware uses something like 80bit (IEEE?) stored as 96bit or 128bit.  But DLPack can't describe it (its an old thing that our `float128` should relaly be `float128_80` or something else than `float128`...).  Something, that will eventually be important if quad precision becomes a real thing.

There are a couple of macros in `numpy/core/src/common/npy_fpmath.h` such as `HAVE_LDOUBLE_IEEE_QUAD_BE`.  I don't know how well they work/are tested, but I guess technically you could check for that to catch the rare cases where we have proper quad precision.

seberg commented:
> How does one check for native dtypes?

You can use `PyDataType_ISNOTSWAPPED`.

eric-wieser commented on numpy/core/src/multiarray/methods.c:
```diff
@@ -2694,6 +2696,159 @@ array_complex(PyArrayObject *self, PyObject *NPY_UNUSED(args))
+    managed->manager_ctx = self;
+    managed->deleter = array_dlpack_deleter;
+
+    PyObject *capsule = PyCapsule_New(managed, NPY_DLPACK_CAPSULE_NAME,
+            array_dlpack_capsule_deleter);
+    if (capsule != NULL) {
+        PyArray_INCREF(self);
+    } else {
+        PyMem_Free(managed);
+        PyMem_Free(managed_shape_strides);
+    }
```
I think sticking to the early return used up until this point makes it easier to follow:
```suggestion
    PyObject *capsule = PyCapsule_New(managed, NPY_DLPACK_CAPSULE_NAME,
            array_dlpack_capsule_deleter);
    if (capsule == NULL) {
        PyMem_Free(managed);
        PyMem_Free(managed_shape_strides);
        return NULL
    }
    
    // the capsule holds a reference
    PyArray_INCREF(self);
```

eric-wieser commented on numpy/core/src/multiarray/methods.c:
```diff
@@ -2694,6 +2696,159 @@ array_complex(PyArrayObject *self, PyObject *NPY_UNUSED(args))
+        return NULL;
+    }
+
+    DLManagedTensor *managed = PyMem_Malloc(sizeof(DLManagedTensor));
+    if (managed == NULL) {
+        PyErr_SetString(PyExc_MemoryError,
+                        "Could not allocate the DLManagedTensor struct.");
+        return NULL;
+    }
+
+    managed->dl_tensor.data = PyArray_DATA(self);
+    managed->dl_tensor.device.device_type = kDLCPU;
+    managed->dl_tensor.device.device_id = 0;
+    managed->dl_tensor.dtype = managed_dtype;
+
+
+    int64_t *managed_shape_strides = PyMem_Malloc(sizeof(int64_t) * ndim * 2);
```
Thoughts on doing this in a single allocation?

hameerabbasi replied:
I believe @seberg had some thoughts in https://github.com/numpy/numpy/pull/19083#discussion_r638066430, I'm not entirely sure how to get the alignment right for the whole thing.

eric-wieser replied:
I believe something like
```
struct dlpack_buffer {
    DLManagedTensor tensor;
    int64 shape_strides[];
}
```
then `PyMem_Malloc(sizeof(dlpack_buffer) + sizeof(int64_t) * ndim * 2)` is the safest way to do it. Perhaps it doesn't matter enough for that awkwardness though.

leofang commented on numpy/core/src/multiarray/methods.c:
```diff
@@ -2694,6 +2696,161 @@ array_complex(PyArrayObject *self, PyObject *NPY_UNUSED(args))
+
+    DLManagedTensor *managed = PyMem_Malloc(sizeof(DLManagedTensor));
+    if (managed == NULL) {
+        PyErr_SetString(PyExc_MemoryError,
+                        "Could not allocate the DLManagedTensor struct.");
```
Maybe this is enough? Do we want to know the detail?
```suggestion
        PyErr_NoMemory();
```

leofang commented on numpy/core/src/multiarray/methods.c:
```diff
@@ -2694,6 +2696,161 @@ array_complex(PyArrayObject *self, PyObject *NPY_UNUSED(args))
+
+    int64_t *managed_shape_strides = PyMem_Malloc(sizeof(int64_t) * ndim * 2);
+    if (managed_shape_strides == NULL) {
+        PyErr_SetString(PyExc_MemoryError,
+                        "Could not allocate the DLManagedTensor struct shape/"
+                        "strides.");
```
ditto
```suggestion
        PyErr_NoMemory();
```

leofang commented on numpy/core/src/multiarray/methods.c:
```diff
@@ -2694,6 +2696,161 @@ array_complex(PyArrayObject *self, PyObject *NPY_UNUSED(args))
+static PyObject *
+array_dlpack_device(PyArrayObject *NPY_UNUSED(self), PyObject *NPY_UNUSED(args))
+{
+    return Py_BuildValue("ii", kDLCPU, 0);
```
Note: This could be `kDLCPUPinned` if [the NumPy array is backed by pinned memory](https://docs.cupy.dev/en/stable/reference/generated/cupyx.empty_pinned.html#cupyx.empty_pinned), but let's get this in first, as we need a mechanism to retrieve this information (maybe #17582?).

hameerabbasi replied:
How do you reliably free such memory when the `np.ndarray` is destroyed? This may be useful when I build `from_dlpack`.

seberg replied:
We don't have to free the memory, the owner is in charge of that (with the deleter)?  I.e. all you do is hold on to the "capsule" as the base object (or at least its `deleter`).

There might be a small issue, that NumPy exposes the `base` object.  DLPack has a "there is only one consumer" contract, but the easiest way would be for NumPy to store the capsule as `arr.base`.  This capsule will *not* be renamed, since it still _owns_ the data (must call the `deleter`).
But such a capsule is not strictly guarded against being consumed a second time.  And that would be wrong, since it would steal ownership from the NumPy array.

Enforcing that would mean wrapping the dlpack capsule into a new object for `arr.base`.  If we need to wrap it into a new object, maybe @eric-wieser's point about a `dlpack` module (possibly even going through the buffer protocol as the final step), becomes more attractive.
(To be clear: It doesn't bother me right now, we could probably transition easy enough.)

About `kDLCPUPinned`: It would be nice if DLPack had an easy way to tell that memory is transparently accessible from the CPU if there are different devices that are.  (A flag would be cute for compatibility if new devices are added, but maybe just a macro?)

hameerabbasi replied:
How about the following: We rename the existing capsule to `"used_dltensor"`, create a new capsule containing the DLPack struct and a new name. This way, the old capsule was consumed and the contract isn't violated, and we're just holding onto the reference of a NumPy "internal" DLPack capsule.

seberg replied:
Yeah, should work well, any intermediate object seems fine.  Or maybe I am misunderstanding and the renamed capsule is always still the owner and must always only be deleted when the data is not needed anymore.  (I.e. in this PR the rename should not prevent cleanup.)  In that case we can just rename and attach the capsule, with the only downside being that `.base` doesn't provide much useful info.

hameerabbasi committed:
ENH, TST: Add the from_dlpack method and test DLPack.

hameerabbasi commented:
This builds fine locally, how does one get past the build warning-as-an-error?

BvB93 commented on numpy/core/_add_newdocs.py:
```diff
@@ -1564,6 +1564,11 @@
 
+add_newdoc('numpy.core.multiarray', 'from_dlpack',
+    """
+    Create a NumPy array from a DLPack struct.
```
```suggestion
    from_dlpack(obj, /)

    Create a NumPy array from an object implementing the ``__dlpack__`` protocol.
```
Three things:
1. The signature was missing.
2. The docstring previously claimed to construct an array from a DLPack struct. 
  From what I understand this is not correct though, as the function takes any object (correctly) implementing the ``__dlpack__`` protocol (which in turn does return a PyCapsule pointing to the DLPack struct in question).
3. It might be useful to add a See Also section with a link to the API-specs of ``__dlpack__``.

BvB93 replied:
Come to think of it, the docstrings of ``ndarray.__dlpack__`` and ``ndarray.__dlpack_device__`` are also missing.

BvB93 commented on numpy/__init__.pyi:
```diff
@@ -1644,6 +1644,10 @@ _ArrayComplex_co = NDArray[Union[bool_, integer[Any], floating[Any], complexfloa
 _ArrayNumber_co = NDArray[Union[bool_, number[Any]]]
 _ArrayTD64_co = NDArray[Union[bool_, integer[Any], timedelta64]]
 
+# `builtins.PyCapsule` unfortunately lacks annotations as of the moment;
+# use `Any` as a stopgap measure
+_PyCapsule = Any
```
Can you also add `from_dlpack` to the stubs?

``` python
_T_contra = TypeVar("_T_contra", contravariant=True)

class _SupportsDLPack(Protocol[_T_contra]):
    def __dlpack__(self, *, stream: _T_contra = ...) -> _PyCapsule: ...

def from_dlpack(__obj: _SupportsDLPack[None]) -> NDArray[Any]: ...
```

eric-wieser commented on numpy/core/src/multiarray/multiarraymodule.c:
```diff
@@ -4228,6 +4230,155 @@ _reload_guard(PyObject *NPY_UNUSED(self)) {
+
+NPY_NO_EXPORT PyObject *
+from_dlpack(PyObject *NPY_UNUSED(self), PyObject *obj) {
+    PyObject *capsule = PyObject_CallMethod(obj, "__dlpack__", NULL);
+    if (capsule == NULL) {
+        return NULL;
+    }
```
This isn't how dunder lookup usually works; does `dlpack` specify these semantics?

The rule for all builtin dunders (`_Py_LookupSpecial`) is roughly `type(obj).__dict__[dunder](obj)`, which bypasses metaclasses and the object dictionary.

eric-wieser commented on numpy/core/src/multiarray/multiarraymodule.c:
```diff
@@ -4228,6 +4230,155 @@ _reload_guard(PyObject *NPY_UNUSED(self)) {
+            strides, data, 0, NULL);
+    if (ret == NULL) {
+        Py_XDECREF(capsule);
+        Py_XDECREF(descr);
```
```suggestion
```
`PyArray_NewFromDescr` owns the reference to `descr` at this point

mattip replied:
fixed

eric-wieser commented on numpy/core/src/multiarray/multiarraymodule.c:
```diff
@@ -4228,6 +4230,155 @@ _reload_guard(PyObject *NPY_UNUSED(self)) {
     Py_RETURN_NONE;
 }
 
+static void array_dlpack_deleter(DLManagedTensor *self)
+{
+    PyArrayObject *array = (PyArrayObject *)self->manager_ctx;
+    // This will also free the strides as it's one allocation.
+    PyMem_Free(self->dl_tensor.shape);
+    PyMem_Free(self);
+    Py_XDECREF(array);
+}
+
+
```
Unused?

hameerabbasi commented on numpy/core/src/multiarray/methods.c:
```diff
@@ -2694,6 +2694,153 @@ array_complex(PyArrayObject *self, PyObject *NPY_UNUSED(args))
+    managed->dl_tensor.strides = managed_strides;
+    managed->dl_tensor.byte_offset = 0;
+    managed->manager_ctx = self;
+    managed->deleter = array_dlpack_deleter;
```
@eric-wieser Used here.

eric-wieser replied:
Yes, but this is a different file with a different definition of `array_dlpack_deleter`

eric-wieser commented on numpy/core/src/common/npy_dlpack.h:
```diff
@@ -0,0 +1,25 @@
+#ifndef NPY_DLPACK_H
+#define NPY_DLPACK_H
+
+#define NPY_DLPACK_CAPSULE_NAME "dltensor"
+#define NPY_DLPACK_USED_CAPSULE_NAME "used_dltensor"
```
```suggestion
/* Part of the DLPack specification */
#define NPY_DLPACK_CAPSULE_NAME "dltensor"
#define NPY_DLPACK_USED_CAPSULE_NAME "used_dltensor"
```

eric-wieser commented on numpy/core/src/common/npy_dlpack.h:
```diff
@@ -0,0 +1,25 @@
+
+#define NPY_DLPACK_CAPSULE_NAME "dltensor"
+#define NPY_DLPACK_USED_CAPSULE_NAME "used_dltensor"
+#define NPY_DLPACK_INTERNAL_CAPSULE_NAME "numpy_dltensor"
```
```suggestion
/* Used internally by numpy for ??? */
#define NPY_DLPACK_INTERNAL_CAPSULE_NAME "numpy_dltensor"
```

eric-wieser commented on numpy/core/src/multiarray/multiarraymodule.c:
```diff
@@ -4228,6 +4230,155 @@ _reload_guard(PyObject *NPY_UNUSED(self)) {
+    PyObject *new_capsule = PyCapsule_New(managed,
+            NPY_DLPACK_INTERNAL_CAPSULE_NAME, array_dlpack_capsule_deleter);
+    if (new_capsule == NULL) {
+        Py_XDECREF(descr);
```
```suggestion
```

eric-wieser commented on numpy/core/src/multiarray/multiarraymodule.c:
```diff
@@ -4228,6 +4230,155 @@ _reload_guard(PyObject *NPY_UNUSED(self)) {
+    }
+
+    if (PyArray_SetBaseObject((PyArrayObject *)ret, new_capsule) < 0) {
+        Py_XDECREF(descr);
```
```suggestion
```

eric-wieser commented on numpy/core/src/multiarray/multiarraymodule.c:
```diff
@@ -4228,6 +4230,155 @@ _reload_guard(PyObject *NPY_UNUSED(self)) {
+    }
+
+    if (PyCapsule_SetName(capsule, NPY_DLPACK_USED_CAPSULE_NAME) < 0) {
+        Py_XDECREF(descr);
+        Py_XDECREF(ret);
+        Py_XDECREF(new_capsule);
```
`new_capsule` should be managed by the array now.
```suggestion
        Py_XDECREF(ret);
```

hameerabbasi committed:
MAINT, BUG: Documentation for DLPack protocol and refcounting bug fixes.

BvB93 commented on numpy/core/_add_newdocs.py:
```diff
@@ -2235,6 +2247,15 @@
     """Array protocol: C-struct side."""))
 
+add_newdoc('numpy.core.multiarray', 'ndarray', ('__dlpack__',
+    """a.__dlpack__(self, /, *, stream=None)
```
```suggestion
    """a.__dlpack__(*, stream=None)
```
Looking the other methods, it seems to be convention to omit the `self` parameter when describing method signatures.
https://github.com/numpy/numpy/blob/0891c5f659a4ddbe952e124410a7d40a3cbd234b/numpy/core/_add_newdocs.py#L2793

leofang commented on numpy/core/src/multiarray/multiarraymodule.c:
```diff
@@ -4229,6 +4231,144 @@ _reload_guard(PyObject *NPY_UNUSED(self)) {
+    }
+
+    const int ndim = managed->dl_tensor.ndim;
+    if (ndim >= NPY_MAXDIMS) {
```
```suggestion
    if (ndim > NPY_MAXDIMS) {
```

leofang commented on numpy/core/src/multiarray/multiarraymodule.c:
```diff
@@ -4229,6 +4231,144 @@ _reload_guard(PyObject *NPY_UNUSED(self)) {
+        return NULL;
+    }
+
+    if (managed->dl_tensor.device.device_type != kDLCPU &&
+            managed->dl_tensor.device.device_type != kDLCUDAHost) {
```
A supported device type list is this: `kDLCPU`, `kDLCUDAHost`, `kDLROCMHost`, `kDLCUDAManaged`. They are all readable/writable on CPU.

leofang replied:
Ah, sorry, the last two are new (https://github.com/dmlc/dlpack/pull/71) and not yet included in the bundled header. Let's add them in a separate PR.

hameerabbasi replied:
I have updated the bundled header, and applied this suggestion.

BvB93 commented on numpy/core/tests/test_dlpack.py:
```diff
@@ -0,0 +1,76 @@
+        np.int8, np.int16, np.int32, np.int64,
+        np.uint8, np.uint16, np.uint32, np.uint64,
+        np.float16, np.float32, np.float64,
+        np.complex64, np.complex64
```
```suggestion
        np.complex64, np.complex128
```
I assume this was meant to be `np.complex128`?

leofang commented on numpy/core/src/multiarray/multiarraymodule.c:
```diff
@@ -4229,6 +4231,144 @@ _reload_guard(PyObject *NPY_UNUSED(self)) {
+    }
+
+    if (PyCapsule_SetName(capsule, NPY_DLPACK_USED_CAPSULE_NAME) < 0) {
+        Py_XDECREF(capsule);
+        Py_XDECREF(ret);
+        return NULL;
```
How about `new_capsule`?

hameerabbasi replied:
`new_capsule` is owned by `ret`, and will be automatically decref'd. See https://github.com/numpy/numpy/pull/19083#discussion_r640708282

eric-wieser replied:
It's owned by `ret` now.

leofang replied:
Ahh yes, I missed it. Thank you.

leofang commented on numpy/core/src/multiarray/multiarraymodule.c:
```diff
@@ -4229,6 +4231,144 @@ _reload_guard(PyObject *NPY_UNUSED(self)) {
+        return NULL;
+    }
+
+    Py_XDECREF(capsule);
```
I am probably being silly: Won't we trigger the deleter from this capsule? Why can't we set the base to be this capsule and instead create a new one?

hameerabbasi replied:
According to the standard, the `"used_dltensor"` has already been consumed, and so it shouldn't own any data. Thus, we create a new capsule, `"numpy_dltensor"` that will be the base.

I have set up the deleter so that it won't delete `"used_dltensor"`, but it will delete `"numpy_dltensor"`.

eric-wieser commented on numpy/core/src/multiarray/multiarraymodule.c:
```diff
@@ -4229,6 +4231,147 @@ _reload_guard(PyObject *NPY_UNUSED(self)) {
+    if (PyArray_SetBaseObject((PyArrayObject *)ret, new_capsule) < 0) {
+        Py_XDECREF(capsule);
+        Py_XDECREF(ret);
+        Py_XDECREF(new_capsule);
```
```suggestion
```
`PyArray_SetBaseObject` steals a reference, even if it fails.

eric-wieser commented on numpy/core/src/common/dlpack/dlpack.h:
```diff
@@ -0,0 +1,199 @@
+/*!
+ *  Copyright (c) 2017 by Contributors
+ * \file dlpack.h
+ * \brief The common header of DLPack.
+ */
```
This should include a commit hash from upstream and probably a URL.

What's the rationale behind everyone having their own copy of this header rather than taking `dlpack` as a dependency?

hameerabbasi replied:
cc @rgommers I think you can better answer this one.

rgommers replied:
> This should include a commit hash from upstream and probably a URL.

That sounds like a good idea.

> What's the rationale behind everyone having their own copy of this header rather than taking dlpack as a dependency?

I think adding a dependency is simply a much larger ask than vendoring a ~200 LoC header. Right now the only runtime dependencies we have are Python itself, and a BLAS library. And build-time dependencies Python/BLAS/Cython/compilers. Adding anything both makes things more fragile, involves extra work for packagers, and introduces coordination issues when the dependency needs updating.

eric-wieser commented on numpy/core/src/multiarray/multiarraymodule.c:
```diff
@@ -4229,6 +4231,146 @@ _reload_guard(PyObject *NPY_UNUSED(self)) {
+    }
+
+    PyObject *new_capsule = PyCapsule_New(managed,
+            NPY_DLPACK_INTERNAL_CAPSULE_NAME, array_dlpack_capsule_deleter);
```
I think it would make sense to use a different deleter here

hameerabbasi committed:
MAINT: Add URL to DLPack GitHub.

eric-wieser commented on numpy/core/src/common/npy_dlpack.h:
```diff
@@ -0,0 +1,30 @@
+// capsule.
+#define NPY_DLPACK_INTERNAL_CAPSULE_NAME "numpy_dltensor"
+
+static void array_dlpack_capsule_deleter(PyObject *self)
+{
+    if (!PyCapsule_IsValid(self, NPY_DLPACK_CAPSULE_NAME) &&
+            !PyCapsule_IsValid(self, NPY_DLPACK_INTERNAL_CAPSULE_NAME)) {
+        if (!PyCapsule_IsValid(self, NPY_DLPACK_USED_CAPSULE_NAME)) {
+            PyErr_SetString(PyExc_RuntimeError, "Invalid capsule name.");
+        }
+        return;
+    }
+    DLManagedTensor *managed = 
+        (DLManagedTensor *)PyCapsule_GetPointer(self, PyCapsule_GetName(self));
+    managed->deleter(managed);
+}
```
If we split this in two, then we can be a little clear about what's dlpack boilerplate and what's our design:
```suggestion
/* This is exactly as mandated by dlpack (and if it isn't, it should be!) */
static void array_dlpack_capsule_deleter(PyObject *self)
{
    if (PyCapsule_IsValid(self, NPY_DLPACK_USED_CAPSULE_NAME)) {
        return;
    }
    DLManagedTensor *managed = 
        (DLManagedTensor *)PyCapsule_GetPointer(self, NPY_DLPACK_CAPSULE_NAME);
    if (managed == NULL) {
        return NULL;
    }
    managed->deleter(managed);
}

/* used internally */
static void array_dlpack_internal_capsule_deleter(PyObject *self)
{
    DLManagedTensor *managed = 
        (DLManagedTensor *)PyCapsule_GetPointer(self, NPY_DLPACK_INTERNAL_CAPSULE_NAME);
    if (managed == NULL) {
        return NULL;
    }
    managed->deleter(managed);
}
```

hameerabbasi replied:
Sigh, this change made the warnings come back, even though both deleters are being used in different parts of the codebase.

eric-wieser replied:
Maybe just move the second one to where it's actually used?

hameerabbasi committed:
MAINT: Split up capsule deleter.

eric-wieser commented on numpy/core/src/common/npy_dlpack.h:
```diff
@@ -0,0 +1,41 @@
+// capsule.
+#define NPY_DLPACK_INTERNAL_CAPSULE_NAME "numpy_dltensor"
+
+/* This is exactly as mandated by dlpack */
```
I put a comment at https://github.com/dmlc/dlpack/issues/51#issuecomment-849794854 suggesting that dlpack suggest this exact implementation.

hameerabbasi committed:
MAINT: Move around code so that there are no more unused warnings.

hameerabbasi committed:
TST: Improve testing coverage for DLPack.

hameerabbasi commented:
What are usually the strides in a 1-element array? It seems things fall apart for the wheel/1-element array case.

eric-wieser commented on numpy/core/src/multiarray/methods.c:
```diff
@@ -2694,6 +2694,166 @@ array_complex(PyArrayObject *self, PyObject *NPY_UNUSED(args))
+    Py_XDECREF(array);
+}
+
+/* This is exactly as mandated by dlpack */
+static void array_dlpack_capsule_deleter(PyObject *self)
+{
+    if (PyCapsule_IsValid(self, NPY_DLPACK_USED_CAPSULE_NAME)) {
+        return;
+    }
+    DLManagedTensor *managed = 
+        (DLManagedTensor *)PyCapsule_GetPointer(self, NPY_DLPACK_CAPSULE_NAME);
+    if (managed == NULL) {
+        return;
+    }
+    managed->deleter(managed);
+}
```
Getting this function right is hard - I've updated my implementation in https://github.com/dmlc/dlpack/issues/51#issuecomment-849794854, but it may need further iteration.

The fact that the glue between dlpack and CPython is hard to get right suggests to me that this should really be handled by a dependency written by the dlpack maintainers, rather than burdening every downstream implementer which writing their own version.

hameerabbasi replied:
We just have to get things right for the cases that we produce. If/when DLPack adds a standard implementation, I'll replace the one here.

leofang replied:
I opened an issue (https://github.com/dmlc/dlpack/issues/74) to discuss about `deleter` and other possible DLPack C APIs.

eric-wieser replied:
@hameerabbasi, you still need the PyErr_Fetch stuff from the comment I link

seberg commented:
> What are usually the strides in a 1-element array? It seems things fall apart for the wheel/1-element array case.

There is no limitation on the strides, it depends on where they come from.  There is a setting that could intentionally mangle it, but I doubt we test it (`NPY_RELAXED_STRIDES_DEBUG`).  The buffer protocol implementation cleans up strides if the array is C- or F-contiguous (C-contiguous you can set them to NULL, but F-contiguous you would have to regenerate them).  (This is partially to work around stricter expectations when contiguity is indicated.)

hameerabbasi commented:
>  There is a setting that could intentionally mangle it, but I doubt we test it (`NPY_RELAXED_STRIDES_DEBUG`).

Thanks, that was it. I'll add handling for this case.

hameerabbasi committed:
BUG: Fix handling of C-contiguous and 1-element arrays.

leofang commented on numpy/core/src/multiarray/methods.c:
```diff
@@ -2694,6 +2694,171 @@ array_complex(PyArrayObject *self, PyObject *NPY_UNUSED(args))
+    if (managed == NULL) {
+        return;
+    }
+    managed->deleter(managed);
```
```suggestion
    if (managed->deleter) { // DLPack header says this can be NULL... 
        managed->deleter(managed);
    }
```

hameerabbasi replied:
Yes, but this deleter is only needed for the capsules NumPy produces, and NumPy never produces capsules with a `NULL` deleter.

leofang replied:
oh yes we export it.

hameerabbasi committed:
BUG, TST: Device bugfix and test __dl_device__.

hameerabbasi commented on numpy/core/src/multiarray/multiarraymodule.c:
```diff
@@ -4228,6 +4230,161 @@ _reload_guard(PyObject *NPY_UNUSED(self)) {
+        }
+    }
+
+    char *data = (char *)managed->dl_tensor.data +
```
@leofang I keep thinking of cases where `data` may become corrupted because of this. For example, take CUDA managed memory as a byte tensor. We do the following

```
x1 = <cuda managed uint8 byte tensor>
x2 = np.from_dlpack(x1)
x3 = x2[1:]  # data no longer 256-byte aligned, numpy doesn't store offset!!!
x4 = cp.from_dlpack(x3) # No longer 256-byte aligned!!!
```

Would this be a problem? If so, the ultimate solution is to check if `arr.base` is a NumPy internal capsule or array, and return a capsule containing the base/offset from that.

The allocation/deallocation happens correctly, just an intermediate data pointer may have the offset integrated and so may not be 256-byte aligned.

hameerabbasi replied:
This is now fixed. The PR will get compute the data/offset from the capsule.

leofang replied:
Thanks, @hameerabbasi. Interestingly I think several implementations missed this (for example, both [CuPy](https://github.com/cupy/cupy/blob/b53e13d15d544604bebca08256d365fb0828e515/cupy/_core/dlpack.pyx#L112) and [PyTorch](https://github.com/pytorch/pytorch/blob/3a66a1cb99d7e1c1fe7abf8f390e177b9183a436/aten/src/ATen/DLConvertor.cpp#L235) set `byte_offset` to 0), though I need to think twice about your b5f3cd3 and ensure I understand it correctly. It would be nice to add a short comment for this.

eric-wieser commented on numpy/core/src/multiarray/methods.c:
```diff
@@ -2694,6 +2694,184 @@ array_complex(PyArrayObject *self, PyObject *NPY_UNUSED(args))
+    managed->dl_tensor.data = PyArray_DATA(self);
+    managed->dl_tensor.device = array_get_dl_device(self);
+    if (PyErr_Occurred()) {
+        return NULL;
```
This leaks `managed`

eric-wieser commented on numpy/core/src/multiarray/methods.c:
```diff
@@ -2694,6 +2694,185 @@ array_complex(PyArrayObject *self, PyObject *NPY_UNUSED(args))
+        DLManagedTensor *managed = PyCapsule_GetPointer(
+                base, NPY_DLPACK_INTERNAL_CAPSULE_NAME);
+        if (managed == NULL) {
+            return ret;
```
This leaves an exception set, which is wrong. This should either return NULL or clear the exception. How can this line even be reached?

hameerabbasi replied:
[`PyCapsule_GetPointer`](https://docs.python.org/3/c-api/capsule.html#c.PyCapsule_GetPointer) can set an exception.

hameerabbasi replied:
And I call `PyErr_Occurred` after if needed.

eric-wieser replied:
> PyCapsule_GetPointer can set an exception.

yes, but the only situation in which it does is the one the outer if protects against - so I'd suggest removing the `if` so that this path is actually taken.

> And I call PyErr_Occurred after if needed.

After what? This is a return statement, and C API functions should do exactly one of return null or set an exception

hameerabbasi replied:
> yes, but the only situation in which it does is the one the outer if protects against - so I'd suggest removing the `if` so that this path is actually taken.

There's no guarantee of this in the documentation that I can find that that's the only case it'll fail.

> You should explain this in the docstring.

I'll explain in the docstring.

eric-wieser replied:
You're right that there is no guarantee of that sort, but you don't need such a guarantee to remove the _outer_ if statement

hameerabbasi replied:
Ah, I see the confusion here -- This is roughly equivalent to the following Python-syntax psuedocode:

```python
if is_numpy_capsule(self.base):  # This is what the outer if is for.
    return self.base.ptr.device
# NumPy arrays by default are on the CPU.
return (kDLCPU, 0)
```

eric-wieser commented on numpy/core/src/multiarray/multiarraymodule.c:
```diff
@@ -4228,6 +4230,161 @@ _reload_guard(PyObject *NPY_UNUSED(self)) {
+/* used internally */
+static void array_dlpack_internal_capsule_deleter(PyObject *self)
+{
+    DLManagedTensor *managed = 
+        (DLManagedTensor *)PyCapsule_GetPointer(self, NPY_DLPACK_INTERNAL_CAPSULE_NAME);
+    if (managed == NULL) {
+        return;
+    }
+    managed->deleter(managed);
```
As above, this is unsafe:

* `deleter` may be null
* `PyCapsule_GetPointer` is unsafe to call if an exception is in flight, so `PyErr_Fetch` is needed first.

eric-wieser replied:
@hameerabbasi, I don't think this one is resolved.

mattip replied:
Why do we need the `array_dlpack_internal_capsule_deleter`? Can't we use the `dlpack_capsule_deleter` instead?

mattip replied:
fixed.

hameerabbasi committed:
MAINT: Robustify dlpack_capsule_deleter and add comments.

hameerabbasi committed:
BUG: Offset not properly stored/computed.

leofang commented:
I am working on the counterpart support for CuPy and I realize we should have this question settled first: For `xp.from_dlpack()`, do we expect to return a NumPy/CuPy array (i.e. just a plain `xp.ndarray`), or an array that's compliant with NEP-47 for NumPy (which CuPy will likely follow) under the new Array API namespace (#18585)? I was assuming the latter, but it looks like the whole review here has been centered around the former?

cc: @kmaehashi @emcastillo @asi1024 @rgommers @asmeurer @hameerabbasi @eric-wieser @seberg @mattip @BvB93

eric-wieser commented on numpy/core/src/multiarray/methods.c:
```diff
@@ -2694,6 +2694,221 @@ array_complex(PyArrayObject *self, PyObject *NPY_UNUSED(args))
+    managed->dl_tensor.dtype = managed_dtype;
+
+
+    int64_t *managed_shape_strides = PyMem_Malloc(sizeof(int64_t) * ndim * 2);
+    if (managed_shape_strides == NULL) {
```
This incorrectly raises an error when `ndim = 0`. Coalescing the allocations would resolve that bug.

mattip replied:
Coalescing which allocations?

dalcinl commented on numpy/core/src/multiarray/multiarraymodule.c:
```diff
@@ -4228,6 +4230,161 @@ _reload_guard(PyObject *NPY_UNUSED(self)) {
+        return NULL;
+    }
+
+    const int ndim = managed->dl_tensor.ndim;
```
Add a check for `ndim < 0`

dalcinl commented on numpy/core/src/multiarray/multiarraymodule.c:
```diff
@@ -4228,6 +4230,161 @@ _reload_guard(PyObject *NPY_UNUSED(self)) {
+    npy_intp shape[NPY_MAXDIMS];
+    npy_intp strides[NPY_MAXDIMS];
+
+    for (int i = 0; i < ndim; ++i) {
```
Before this loop, add a check for `shape == NULL` if `ndim > 0`.

seberg replied:
Thanks for the comments, but at least I expect that the exporter ensures that this is not clearly malformed (since this is C-API and not Python input/API).  For half of these, they are actually also checked during the array creation.

dalcinl replied:
I just did my best to find potential problems. Pleases resolve my comments if you consider any of them a non-issue.

dalcinl commented on numpy/core/src/multiarray/multiarraymodule.c:
```diff
@@ -4228,6 +4230,161 @@ _reload_guard(PyObject *NPY_UNUSED(self)) {
+    npy_intp strides[NPY_MAXDIMS];
+
+    for (int i = 0; i < ndim; ++i) {
+        shape[i] = managed->dl_tensor.shape[i];
```
Add a check for `shape[i] < 0`.

dalcinl commented on numpy/core/src/multiarray/multiarraymodule.c:
```diff
@@ -4228,6 +4230,161 @@ _reload_guard(PyObject *NPY_UNUSED(self)) {
+        // DLPack has elements as stride units, NumPy has bytes.
+        if (managed->dl_tensor.strides != NULL)
+        {
+            strides[i] = managed->dl_tensor.strides[i] * itemsize;
```
I'm not sure whether `strides[i] < 0` would be valid or not. If not, please add a check.

dalcinl commented on numpy/core/src/multiarray/multiarraymodule.c:
```diff
@@ -4228,6 +4230,161 @@ _reload_guard(PyObject *NPY_UNUSED(self)) {
+        }
+    }
+
+    char *data = (char *)managed->dl_tensor.data +
```
Add a check for `data == NULL` if `product(shape) > 0`.

rgommers commented:
>  For `xp.from_dlpack()`, do we expect to return a NumPy/CuPy array (i.e. just a plain `xp.ndarray`), or an array that's compliant with NEP-47 for NumPy (which CuPy will likely follow) under the new Array API namespace (#18585)? I was assuming the latter, but it looks like the whole review here has been centered around the former?

The question is a little ambiguous (use of `xp`). The intent is:
1. `numpy.from_dlpack` returns a `numpy.ndarray`
2. `numpy.array_api.from_dlpack` returns the new, array API compliant, array instance

DLPack is useful standalone, independent of NEP 47. This PR therefore adds DLPack support to NumPy in its main namespace (so only (1) above).

leofang commented:
> DLPack is useful standalone, independent of NEP 47.

Ah OK, thanks @rgommers, it makes sense although I didn't anticipate this. We should add a clause to https://data-apis.org/array-api/latest/design_topics/data_interchange.html to clarify this. The way I read it as in its current form, I would think `from_dlpack()` returns an array-API-compliant object, not any random object, but obviously we do not want to impose this strict requirement.

rgommers commented:
> The way I read it as in its current form, I would think `from_dlpack()` returns an array-API-compliant object, not any random object,

It should do that! Assuming the namespace you call it in conforms to the array API:) In this way it's no different from `numpy.zeros` vs. `numpy.array_api.zeros`.

leofang commented:
Right, I was just trying to make sure we are allowed to add `from_dlpack()` to non-compliant namespaces like `numpy` and `cupy`, and get a non-complaint object back 🙂

seberg commented on numpy/core/src/multiarray/methods.c:
```diff
@@ -2694,6 +2694,221 @@ array_complex(PyArrayObject *self, PyObject *NPY_UNUSED(args))
+        // We can't be sure that the dtype is
+        // IEEE or padded.
+        if (itemsize > 8) {
+            PyErr_SetString(PyExc_TypeError, "DLPack only supports IEEE "
+                    "floating point types without padding.");
```
NIT:
```suggestion
            PyErr_SetString(PyExc_TypeError,
                    "DLPack only supports IEEE floating point types without "
                    "padding.");
```

seberg commented on numpy/core/tests/test_dlpack.py:
```diff
@@ -0,0 +1,103 @@
+        y2 = x[:, 0]
+        assert_array_equal(y2, np.from_dlpack(y2))
+
+        y3 = x[1, :]
```
Could parametrize, but its fine.  Also `::2` is a nice, since then you got a 2-D non-contiguous one.

seberg commented on numpy/core/src/multiarray/methods.c:
```diff
@@ -2694,6 +2694,221 @@ array_complex(PyArrayObject *self, PyObject *NPY_UNUSED(args))
+        return NULL;
+    }
+
+    DLDevice device = array_get_dl_device(self);
+    if (PyErr_Occurred()) {
+        return NULL;
+    }
+    char *data = array_get_dl_data(self);
+    if (data == NULL) {
+        return NULL;
+    }
```
(if we do, it would be nice to add a test, since I think calling `from_dlpack` twice on the same array will do the trick?)

seberg reviewed (commented):
I personally would like to just create a new `dlpack_impl.c` and `dlpack_impl.h` in `multiarray`.  Then export the `__dlpack__` and `from_dlpack` there include it into `methods.c` and `multiarray.c` and keep all of DLPack in one place.  Especially `multiarray.c` is just too unnecessarily long, and its easy to go from the method list/struct to the actual definition in the other file.

BvB93 commented on numpy/__init__.pyi:
```diff
@@ -3654,3 +3661,8 @@ class broadcast:
     def reset(self) -> None: ...
+
+class _SupportsDLPack(Protocol[_T_contra]):
+    def __dlpack__(self, *, stream: Optional[int] = ...) -> _PyCapsule: ...
```
```suggestion
    def __dlpack__(self, *, stream: None | _T_contra = ...) -> _PyCapsule: ...
```
The `|` operator is just a more compact shortcut for `Union` or `Optional`,
but more importantly: the typevar is missing from the argument type.

mattip replied:
The signature changed to 
```
def __dlpack__(self: NDArray[number[Any]], *, stream: None = ...) -> _PyCapsule: ...
```

BvB93 replied:
My comment applied (and still does) to the `_SupportsDLPack` protocol, used by the `from_dlpack` function for identifying objects implementing `__dlpack__`.

The `ndarray.__dlpack__` method itself does look perfectly fine.

mattip committed:
move dlpack functions to a new file

mattip committed:
BUG: fixes from review

hameerabbasi commented:
This fell off my radar. Thanks for the fix, @mattip. I can take over if any more changes need to be done.

hameerabbasi commented on numpy/core/tests/test_dlpack.py:
```diff
@@ -0,0 +1,103 @@
+        y4 = x[1]
+        assert_array_equal(y4, np.from_dlpack(y4))
+
+        y5 = np.diagonal(x)
```
@seberg In reference to https://github.com/numpy/numpy/pull/19083/files#r646902645, here's a 2-D non-contiguous one.

mattip commented on numpy/core/src/multiarray/multiarraymodule.c:
```diff
@@ -4228,6 +4230,161 @@ _reload_guard(PyObject *NPY_UNUSED(self)) {
+
+    if (managed->dl_tensor.dtype.lanes != 1) {
+        PyErr_SetString(PyExc_RuntimeError,
+                "Unsupported lanes in DLTensor dtype.");
```
Let's leave this for a future PR when we understand better what the error message should be

mattip commented on numpy/core/src/multiarray/multiarraymodule.c:
```diff
@@ -4228,6 +4230,161 @@ _reload_guard(PyObject *NPY_UNUSED(self)) {
+    }
+
+    PyArray_Descr *descr = PyArray_DescrFromType(typenum);
+    if (descr == NULL) {
```
It cannot fail on a memory error?

mattip commented on numpy/core/src/multiarray/multiarraymodule.c:
```diff
@@ -4228,6 +4230,161 @@ _reload_guard(PyObject *NPY_UNUSED(self)) {
+            device_type != kDLCUDAManaged) {
+        PyErr_SetString(PyExc_RuntimeError,
+                "Unsupported device in DLTensor.");
+        Py_XDECREF(capsule);
```
done

mattip commented on numpy/core/src/multiarray/multiarraymodule.c:
```diff
@@ -4228,6 +4230,161 @@ _reload_guard(PyObject *NPY_UNUSED(self)) {
+        return NULL;
+    }
+
+    if (PyCapsule_SetName(capsule, NPY_DLPACK_USED_CAPSULE_NAME) < 0) {
```
There may be some subtle race condition somewhere here.

mattip commented:
@hameerabbasi I don't know if this was actually completed?



>>    There is a setting that could intentionally mangle it, but I doubt we test it (NPY_RELAXED_STRIDES_DEBUG).

> Thanks, that was it. I'll add handling for this case.

mattip commented on numpy/core/src/multiarray/methods.c:
```diff
@@ -2694,6 +2694,221 @@ array_complex(PyArrayObject *self, PyObject *NPY_UNUSED(args))
+        PyErr_SetString(PyExc_TypeError, "DLPack only supports native "
+                    "byte swapping.");
+            return NULL;
+    }
```
This is mentioned at the top of the PR and is still open

mattip commented on numpy/core/src/multiarray/methods.c:
```diff
@@ -2694,6 +2694,221 @@ array_complex(PyArrayObject *self, PyObject *NPY_UNUSED(args))
+        managed_dtype.code = kDLInt;
+    } else if (PyDataType_ISUNSIGNED(dtype)) {
+        managed_dtype.code = kDLUInt;
+    } else if (PyDataType_ISFLOAT(dtype)) {
```
done

mattip commented:
Other than the read-write question, and the comment about export rejection at the top, I think this is ready for a final round of reviews. 

I think there could be some refactoring around the two deallocators, and we may want to check that we properly raise on pathologically wrong dlpack input, but we can do that in follow-on PRs.

seberg commented on numpy/core/src/multiarray/dlpack.c:
```diff
@@ -0,0 +1,406 @@
+    for (int i = 0; i < ndim; ++i) {
+        shape[i] = managed->dl_tensor.shape[i];
+        // DLPack has elements as stride units, NumPy has bytes.
+        if (managed->dl_tensor.strides != NULL)
+        {
```
```suggestion
        if (managed->dl_tensor.strides != NULL) {
```

seberg reviewed (commented):
Thanks, I did not review the cleanup logic in detail, but I think it was pretty much settled and the main reason for stalling was the under-specification of DLPack.  So basically, this is just a bunch of nits, and it seems fine to defer further cleanup.

Didn't we say we would rename to `_from_dlpack`?  Probably just a commit missing/not pushed?  For myself, I still think `__dlpack__` in its current form is a mistake so I am not in favor of about adding a new symbol for it.  But fair enough if you disagree with me there.

> I think there could be some refactoring around the two deallocators, and we may want to check that we properly raise on pathologically wrong dlpack input, but we can do that in follow-on PRs.

Yes, it seems overly complicated, but it is how the `__dlpack__` is specified IIRC.  It is to protect  from importing the same capsule twice. (which could be fine if the capsule owned the data, but DLPack expects us to transfer ownership away from the capsule).
Probably the reason is that DLPack is designed for the importer to work with the DLPack struct directly in C++ so that seems easier to side-line the capsule as a potential owner.  (Maybe there is even some argument in favor to ensure immediate clean-up not depending on reference counting :shrug:.)

Of course apparently pytorch simply doesn't do *any* ownership management at all: I guess it just hopes the exporting object will outlive the tensor (and did not make a copy, which is _maybe_ invalid?)… https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/DLConvertor.cpp#L245

@rgommers, there is still the under-specification about unaligned, or less aligned data (unaligned data is not yet rejected, presumably it has to be) and no rejection of exporting readonly ("truly readonly"?) data which would be good to clear up before it can be fixed in the code.  That can just be a quick executive decisions, but I feel it should be reflected in a PR to `dlpack.h` and/or data-apis.

leofang commented:
> > I think there could be some refactoring around the two deallocators, and we may want to check that we properly raise on pathologically wrong dlpack input, but we can do that in follow-on PRs.
> 
> Yes, it seems overly complicated, but it is how the `__dlpack__` is specified IIRC. It is to protect from importing the same capsule twice. (which could be fine if the capsule owned the data, but DLPack expects us to transfer ownership away from the capsule). Probably the reason is that DLPack is designed for the importer to work with the DLPack struct directly in C++ so that seems easier to side-line the capsule as a potential owner. (Maybe there is even some argument in favor to ensure immediate clean-up not depending on reference counting 🤷.)
> 
> Of course apparently pytorch simply doesn't do _any_ ownership management at all: I guess it just hopes the exporting object will outlive the tensor (and did not make a copy, which is _maybe_ invalid?)… https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/DLConvertor.cpp#L245

I've been puzzled about the two-capsule approach adopted here. It seems way too complicated to me.

In CuPy, we simply let the capsule own the CuPy array (ref count +1), and then expect the consumer library to hold the incoming capsule, so as long as the capsule is alive so will be the underlying array.
https://github.com/cupy/cupy/blob/afec5a86361cf7650f261c62a331c7f2e5a38c1b/cupy/_core/dlpack.pyx#L81-L87
I think the PyTorch lines you linked to also apply the same approach.

seberg commented:
@leofang maybe your version is "correct", and renaming is supposed to be sufficient:

> PyCapsule name to "used_dltensor" , and call the deleter of the DLPackManagedTensor when it no longer needs the data.

It says to call the deleter of the `DLPackManagedTensor`, and *not* just delete capsule.  So presumably renaming the capsule means we "steal" the reference to `DLPackManagedTensor`.  But we *must* rename the capsule, because otherwise someone else CAN still steal the reference (which would disown us).

As for pytorch, it is obviously incorrect, since it doesn't even rename the capsule.

seberg commented:
In other words... cleanup is *also* under-specified, since the (IMO correct implementation here) should never get cleaned up if imported into cupy.

leofang commented:
> As for pytorch, it is obviously incorrect, since it doesn't even rename the capsule.

I think it's done in this line:
https://github.com/pytorch/pytorch/blob/8854817f448fa23485b24303d4665b2db71feba5/torch/csrc/utils/tensor_new.cpp#L1062
Right after `at::fromDLPack` (that contains the lines you quoted) is called 🙂

seberg commented:
> Right after at::fromDLPack (that contains the lines you quoted) is called slightly_smiling_face

Ah, thanks! I guess I misread the C++ code and thought it was a shorter wrapper.  In that case, maybe it is all clear: cupy also wraps the data into its own `DLPackMemory` so effectively has "two capsules" also.  Error paths may still be iffy, but lets assume they are fine.

leofang commented:
Yup if you interpret this way it's ok to call it two-capsule one, though to me it is fundamentally different for code readers. One literally uses two capsules whereas the other has a clear hierarchal structure (ndarray <- DLPackMemory <- capsule).

On the renaming thing: To me it is largely a legacy practice. It used to be that a capsule is leaked into the user space and so people had to invent a way to prevent it from being reused. But with the advent of the protocol the capsule is exchanged between two libs during handshake, and there's no way to leak it. I'd say it's also fine to keep the capsule name, but since no one complains we'll stick to what's decided.

seberg commented:
> I'd say it's also fine to keep the capsule name as is, but since no one complains we'll stick to what's decided.

Well, it prevents pure-Python code from being able to break the interpreter.  That is good.

> clear hierarchal structure (`ndarray <- DLPackMemory <- capsule`)

I guess you can see it that way, but I see no reason for `DLPackMemory` to hold on to the original capsule!  At least unless there is some reasons due to stream handling.

hameerabbasi commented:
For background, I may not get to this before the weekend, but I'll respond to everything then -- And push changes to what's already settled.

mattip committed:
Updates

Co-authored-by: Sebastian Berg <sebastian@sipsolutions.net>
Co-authored-by: Bas van Beek <43369155+BvB93@users.noreply.github.com>

mattip committed:
change from_dlpack to _dlpack, remove unused header

mattip commented:
Thanks @hameerabbasi. It is a bit difficult to scan through to see what is open, so I am repeating it here. Some of these can wait for a future PR, in which case we should open a tracking issue after merging this:

- [ ] Clarifiy alignment requirements (and use of byte_offset). After such a PR exists, we can review the current rules when to reject export. (if the rejection is not guaranteed to be fixed by arr.copy() we may have a problem.) This is expressed in the line `managed->dl_tensor.byte_offset = PyArray_BYTES(self) - data;`
- [ ] Resolve the rules around conversion of a read-only NumPy array to a dlpack PyCapsule. For this PR, such a conversion should fail. We can relax this restriction going forward.
- [x] dmlc/dlpack#34 about versioning DLPack is still open, with the general consensus that it should be in the DLPack C struct and not only in Python.
- [ ] "Interestingly, if we could get builtins.PyCapsule in typeshed annotated as a parameterizable type,
then it would in principle be possible for static type checkers to read, e.g., the underlying shape and dtype. The only thing that users or libraries would have to do here is declare the necessary annotations. Perhaps something to consider for the future?" (@BvB93). Leaving this here in case we want to subclass PyCapsule to add the methods in a future PR.
- [ ] Add tests for pathologically wrong DLPack importing: structs with `ndim <0`, strides that don't make sense, ...
- [ ] "[What happens on import] when ndim = 0. Coalescing the allocations would resolve that bug." (@eric-wieser). I am not sure we properly addressed this, a test is needed.
- [ ] Do we need a check on `PyArray_DescrFromType(typenum)` since "This can't actually fail. But its obscure enough that I think its fine to keep." (@seberg). Can we reject this suggestion?
- [ ] What happens if `PyCapsule_SetName()` fails, do we handle the error correctly?
- [ ] Add a test for calling `_from_dlpack` twice including deallocation
- [ ] `array_get_dl_data(PyArrayObject *self) { PyObject *base = PyArray_BASE(self); ...` "Should this be recursive/iterative in case we have multiple nested arrays (does not happen often, but can happen in subclassing context)?" (@seberg)

In my opinion, all of this can be done in future PRs, we can put this in as-is

leofang commented:
> > clear hierarchal structure (`ndarray <- DLPackMemory <- capsule`)
> 
> I guess you can see it that way, but I see no reason for `DLPackMemory` to hold on to the original capsule! At least unless there is some reasons due to stream handling.

No, it has nothing to do with stream handling (which is done at the Python level during handshaking), it's about keeping the ref count of the underlying resource. Conceptually we could definitely use a capsule to back an ndarray, in practice though `cupy.ndarray` expects a few extra methods from its `data` member to return device, size, pointer, etc, hence wrapping the capsule with a very thin layer (`DLPackMemory`).

seberg commented:
Not really the right place to continue discussing.  But no: The capsule owns the `DLManagedTensor` which owns its own memory and is not reference counted (there is exactly one reference that cannot be shared).  Renaming the capsule steals ownership from it (and you must rename it).
At that point, the capsule does not own anything (check the capsule `destructor` it does nothing after renaming).  So you simply can delete the `cdef object dltensor` and `self.dltensor = dltensor` line.  The only tricky thing is memory handling (i.e. renaming the capsule at the right time), and I think `cupy` should move that directly after the line getting the pointer from the capsule (at that point `DLPackMemory` has ownership).  That looks like a bug, for example if you are confronted with an unknown dtype.

mattip committed:
make a.__dlpack__() fail if a is readonly

mattip committed:
add release note, error out if offset is used

mattip commented on numpy/core/src/multiarray/dlpack.c:
```diff
@@ -206,6 +206,12 @@ array_dlpack(PyArrayObject *self,
+                        "Offsets not clearly supported by this "
+                        "version of DLPack.");
+        return NULL;
+    }
```
This code path can only be hit if we are re-wrapping a dlpack capsule with an offset. It is currently not tested, which should be done in a follow-on PR.

seberg replied:
Yeah, don't worry about it.  This is nonsense anyway, I deleted the whole thing (the logic was probably incomplete anyway) and replaced it with the following comment:
```
    /*
     * Note: the `dlpack.h` header suggests/standardizes that `data` must be
     * 256-byte aligned.  We ignore this intentionally, because `__dlpack__`
     * standardizes that `byte_offset` must be 0 (for now) to not break pytorch:
     * https://github.com/data-apis/array-api/issues/293#issuecomment-964111413
     *
     * We further assume that exporting fully unaligned data is OK even without
     * `byte_offset` since the standard does not reject it.
     * Presumably, pytorch will support importing `byte_offset != 0` and NumPy
     * can choose to use it starting about 2023.  At that point, it may be
     * that NumPy MUST use `byte_offset` to adhere to the standard (as
     * specified in the header)!
     */
```

seberg committed:
MAINT: Simplify `byte_offset` handling in dlpack.h and add comment

If `byte_offset = 0` is forced anyway, there is no point in trying
to preserve a previous `data` information from the capsule.
(And probably it should have used a base array also, and not just
a base DLPack capsule, anyway.)

seberg commented:
I will put this in once the tests pass, but will then open a few issues about dlpack, with pretty harmless bugs or general followups.  The release notes get reviewed before the release anyway.

Thanks everyone and especially Matti for doing the follow-ups.

rgommers commented:
🎉 thanks @mattip, @hameerabbasi, @seberg, @leofang & all other reviewers!

leofang commented:
Silly question: Can someone remind me why we renamed it to `np._from_dlpack()` (prefixed with an underscore)?

seberg commented:
This was pushed in a ~1 week before branching with not quite settled things in dlpack itself (as I remember).  Adding the underscore was the way to not having to discuss all of that in a short time frame.

leofang commented:
I see, thanks for the context @seberg. Sounds like a decision made offline (which I am fine with). Is there any plan to make it a public API (by removing the underscore)?

seberg commented:
Yes, we can discuss this obviously.  It may have been discussed in a meeting, in which case it is publically noted in the `numpy/archive` repo.  I feel the PR had the underscore initially, because the first step was to make it available for the `_array_api` namespace.

mattip commented:
The only mention of dlpack in the docs is in the release notes, so maybe we could discuss making it public together with some documentation.

leofang commented:
Thanks, guys. I created an issue to track this change: https://github.com/numpy/numpy/issues/20743.

vadimkantorov commented:
Does `numpy._from_dlpack` support construction from a raw capsule?

E.g. PyTorch supports both raw capsule input and an object having `__dlpack__` instance method:
https://pytorch.org/docs/stable/generated/torch.from_dlpack.html:
`If ext_tensor is a tensor (or ndarray) object, it must support the __dlpack__ protocol (i.e., have a ext_tensor.__dlpack__ method). Otherwise ext_tensor may be a DLPack capsule, which is an opaque PyCapsule instance, typically produced by a to_dlpack function or method.`

---
Original issue: 
- https://github.com/dmlc/dlpack/issues/55

My original DLpack python/numpy bindings:
- https://github.com/vadimkantorov/pydlpack/blob/master/dlpack.py

vadimkantorov commented:
@hameerabbasi PyTorch's https://pytorch.org/docs/stable/dlpack.html#torch.utils.dlpack.to_dlpack actually returns an opaque Capsule object... What would be nice is having an example of producing a DLPack struct from C function (via ctypes) or PyCapsule from PyTorch - and then consuming it from NumPy - to showcase the interop situations

I did some examples like this in https://github.com/vadimkantorov/pydlpack and https://github.com/vadimkantorov/readaudio

rgommers commented:
@vadimkantorov `torch.utils.dlpack.to_dlpack` is legacy, just don't use it. PyTorch tensors [have a `__dlpack__` method](https://github.com/pytorch/pytorch/blob/e0d4c43ad104fd72fa2ae412de563e91ebfde913/torch/_tensor.py#L1676), so you can simply use `np.from_dlpack(a_tensor)`.

vadimkantorov commented:
The usecase where DLPack is used for marshalling an array from C land (function can be bound with DLPack) into a NumPy tensor is also very useful: plain C functions with DLPack structures is much simpler to compile / manage than building torch/numpy extensions

In my example I used DLPack-producing C functions doing audio decoding, and then ingesting such a DLPack ctypes-bound structure in NumPy. Does NumPy also provide such ctypes binding types for DLPack structures?

rgommers commented:
It doesn't. But if you already have a raw capsule, it should be quite easy to wrap it in a thin pure Python class with `__dlpack__` and `__dlpack_device__` methods - that should be all you need.

vadimkantorov commented:
For some reason (maybe inspired with original PyTorch support), Triton also returns capsule objects: 
https://docs.nvidia.com/deeplearning/triton-inference-server/user-guide/docs/python_backend/README.html#interoperability-and-gpu-support
